engine/src/main/java/org/hibernate/search/engine/backend/work/execution/OperationSubmitter.java

```diff
@@ -8,52 +8,128 @@
 
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.RejectedExecutionException;
+import java.util.function.Consumer;
+import java.util.function.Function;
 
+import org.hibernate.search.engine.common.execution.spi.SimpleScheduledExecutor;
 import org.hibernate.search.util.common.annotation.Incubating;
 
 
 /**
- * Interface defining how operation should be submitted to the queue.
+ * Interface defining how operation should be submitted to the queue or executor.
+ * <p>
+ * <strong>WARNING:</strong> while this type is API, because instances should be manipulated by users,
+ * all of its methods are considered SPIs and therefore should never be called directly by users.
+ * In short, users are only expected to get instances of this type from an API and pass it to another API.
+ * <p>
+ * Currently supported implementations:
+ * <ul>
+ *     <li>{@link #blocking()}</li>
+ *     <li>{@link #rejectedExecutionException()}</li>
+ *     <li>{@link #offloading(Consumer)}</li>
+ * </ul>
  */
 @Incubating
-public enum OperationSubmitter {
+public abstract class OperationSubmitter {
+
+	private static final OperationSubmitter BLOCKING = new BlockingOperationSubmitter();
+	private static final OperationSubmitter REJECTED_EXECUTION_EXCEPTION = new RejectedExecutionExceptionOperationSubmitter();
+
+	private OperationSubmitter() {
+	}
+
+	/**
+	 * Defines how an element will be submitted to the queue.
+	 * Depending on the implementation might throw {@link RejectedExecutionException} or offload the submit operation to a provided executor.
+	 */
+	public abstract <T> void submitToQueue(BlockingQueue<? super T> queue, T element,
+			Consumer<? super T> blockingRetryProducer) throws InterruptedException;
+
+	/**
+	 * Defines how an element will be submitted to the executor.
+	 * Depending on the implementation might throw {@link RejectedExecutionException} or offload the submit operation to an offload executor.
+	 */
+	public abstract <T extends Runnable> void submitToExecutor(SimpleScheduledExecutor executor, T element,
+			Function<? super T, Runnable> blockingRetryProducer) throws InterruptedException;
 
 	/**
 	 * When using this submitter, dding a new element will block the thread when the underlying
 	 * queue is a {@link java.util.concurrent.BlockingQueue} and it is at its maximum capacity, until some elements
 	 * are removed from the queue
 	 */
-	BLOCKING {
-		@Override
-		public <T> void submitToQueue(BlockingQueue<T> queue, T element) throws InterruptedException {
-			queue.put( element );
-		}
-	},
+	public static OperationSubmitter blocking() {
+		return BLOCKING;
+	}
+
 	/**
 	 * When using this submitter adding a new element will cause a {@link RejectedExecutionException} when the underlying
 	 * queue is a {@link java.util.concurrent.BlockingQueue} and it is at its maximum capacity.
 	 */
-	REJECTED_EXECUTION_EXCEPTION {
+	public static OperationSubmitter rejectedExecutionException() {
```

Comment:
Just something to consider, if you think it makes sense: now that you have `blocking()` on one side and `offloading()` on the other, naming would feel more consistent if this one was called simply `rejecting()`.